### PR TITLE
Support for DISTINCT ON (columns)

### LINKED ISSUE #25404: Support for DISTINCT ON (columns)
For people who get used to Postgres.

**Describe the solution you'd like**
`SELECT DISTINCT ON (a, b, ...) a, b, c, ...`
is equivalent to
`SELECT a, b, c... LIMIT 1 BY a, b, ...`

### PR DESCRIPTION
**Pull request by TszKitLo40**
Signed-off-by: Zijie Lu <wslzj40@gmail.com>

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add support `DISTINCT ON (columns)` expression, close https://github.com/ClickHouse/ClickHouse/issues/25404

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

cc @alexey-milovidov

### CODE REVIEW

**Comment by TszKitLo40** [comment]
PTAL @alexey-milovidov

File: src/Parsers/ParserSelectQuery.cpp [hunk 1]
```diff
@@ -264,6 +277,9 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
             if (limit_with_ties_occured)
                 throw Exception("Can not use WITH TIES alongside LIMIT BY", ErrorCodes::LIMIT_BY_WITH_TIES_IS_NOT_SUPPORTED);
 
+            if (has_distinct_on)
```
Comment (vdimir) [inline src/Parsers/ParserSelectQuery.cpp:280]: I suppose we can check if `limit_by_length` (or `limit_by_expression_list`) already initialized and do not need flag

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS t1;
+
+CREATE TABLE t1 (`a` UInt32, `b` UInt32, `c` UInt32 ) ENGINE = Memory;
+INSERT INTO t1 VALUES (1, 1, 1), (1, 1, 2), (2, 2, 2), (1, 2, 2);
+
+SELECT DISTINCT ON (a, b) a, b, c FROM t1;
+
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:7]: This query not ok
```
SELECT DISTINCT ON(a, b) * FROM AS t1
```

```
Received exception from server (version 21.7.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'b' 'FROM' 'a' 't1' while processing query: 'SELECT t1 LIMIT 1 BY (a, b) * FROM', required columns: 't1' 'a' 'FROM' 'b'.
```

Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:7]: This query works, but looks wired, maybe should not allow such syntax without brackets?
```
:) SELECT DISTINCT ON a, b a, b FROM t1;

SELECT
    a,
    b
FROM t1
LIMIT 1 BY
    a,
    b

Query id: 5ebad59a-911a-4932-beae-20ef75a1a157

┌─a─┬─b─┐
│ 1 │ 1 │
│ 2 │ 2 │
│ 1 │ 2 │
└───┴───┘
```

Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:7]: Also not sure should we forbid `DISTINCT ON` and `DISTINCT`
```
:) SELECT DISTINCT ON  (a, b) DISTINCT a, b FROM t1;

SELECT DISTINCT
    a,
    b
FROM t1
LIMIT 1 BY (a, b)

Query id: f57b59e3-3baf-4874-b00a-6e452b21d439

┌─a─┬─b─┐
│ 1 │ 1 │
│ 2 │ 2 │
│ 1 │ 2 │
└───┴───┘
```

Maybe it's ok

File: src/Parsers/ParserSelectQuery.cpp [hunk 3]
```diff
@@ -94,6 +97,8 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
         }
     }
 
+    bool has_distinct_on = false;
+
     /// SELECT [ALL/DISTINCT] [TOP N [WITH TIES]] expr list
```
Comment (vdimir) [inline src/Parsers/ParserSelectQuery.cpp:102]:
```suggestion
    /// SELECT [DISTINCT ON expr] [ALL/DISTINCT] [TOP N [WITH TIES]] expr list
```

File: src/Common/ErrorCodes.cpp [hunk 4]
```diff
@@ -555,6 +555,7 @@
     M(585, CANNOT_PARSE_YAML) \
     M(586, CANNOT_CREATE_FILE) \
     M(587, CONCURRENT_ACCESS_NOT_SUPPORTED) \
+    M(588, DISTINCT_ON_AND_LIMIT_BY_TOGETHER) \
```
Comment (vdimir) [inline src/Common/ErrorCodes.cpp:558]: I see that we have separate specific error codes for similar cases. But maybe better to reuse some to not introduce new one, but I'm not sure.

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS t1;
+
+CREATE TABLE t1 (`a` UInt32, `b` UInt32, `c` UInt32 ) ENGINE = Memory;
+INSERT INTO t1 VALUES (1, 1, 1), (1, 1, 2), (2, 2, 2), (1, 2, 2);
+
+SELECT DISTINCT ON (a, b) a, b, c FROM t1;
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:6]: Also it's good to add test to `DISTINCT ON` and `LIMIT BY` to check that correct error returned. To check error code add `-- { serverError <exprecred_code> }` like this

https://github.com/ClickHouse/ClickHouse/blob/9f5cd35a6963cc556a51218b46b0754dcac7306a/tests/queries/0_stateless/01601_accurate_cast.sql#L1

**Review by vdimir — CHANGES_REQUESTED** [review]
Thank you for your work!

Some comments to discuss

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 2 above)
Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:7]: Oh, I aimed to have 3 separate conversations for this cases, but it all pinned to one line of code 😕

Comment (TszKitLo40) [reply tests/queries/0_stateless/01917_distinct_on.sql:7]: For the first case, I don't know how to deal with it.
For the second case, do you mean we should add a line `if (open_bracket.ignore(pos, expected))` before L114.
For the third case, it seems that we can allow both `DISTINCT ON` and `DISTINCT`?

File: src/Common/ErrorCodes.cpp (hunk 4 above)
Comment (TszKitLo40) [reply src/Common/ErrorCodes.cpp:558]: But I can't find this error code. Can u please help me figure it out?

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 5 above)
Comment (TszKitLo40) [reply tests/queries/0_stateless/01917_distinct_on.sql:6]: I have added a test.

Comment (TszKitLo40) [reply tests/queries/0_stateless/01917_distinct_on.sql:6]: But the test failed which returned `Code: 590. DB::Exception: Can not use DISTINCT ON alongside LIMIT BY: 
Expected server error '590'.`

File: src/Common/ErrorCodes.cpp (hunk 4 above)
Comment (vdimir) [reply src/Common/ErrorCodes.cpp]: `62` doesn't look suitable, actually I want to find suitable existing code, but seems there isn't. Ok, let's leave as is, I'll change it if I'll find solution or leave current one.

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 5 above)
Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:6]: Exception on server returned as server error (if it's not `LOGICAL_ERROR`). Failed test case looks weird I'll check it.

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 2 above)
Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:7]: > For the first case, I don't know how to deal with it.

Ok, I'll try to figure it out and will suggest to you.

> For the second case, do you mean we should add a line if (open_bracket.ignore(pos, expected)) before L114.

Also need investigation.

> For the third case, it seems that we can allow both DISTINCT ON and DISTINCT?

I'm not sure. If we perform `DISTINCT ON` on any columns subset it implies `DISTINCT`, so it looks redundant. Maybe there's some edge case that I don't notice?

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 5 above)
Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:6]: Because it's `clientError`, not `serverError`

File: src/Common/ErrorCodes.cpp (hunk 4 above)
Comment (vdimir) [reply src/Common/ErrorCodes.cpp]: I've changed to `UNSUPPORTED_METHOD`

File: tests/queries/0_stateless/01917_distinct_on.sql (hunk 2 above)
Comment (vdimir) [reply tests/queries/0_stateless/01917_distinct_on.sql:9]: > For the first case, I don't know how to deal with it.

Oh, sorry, I pasted parsed query. Original query was `SELECT DISTINCT ON (a, b) * FROM t1` but it wasn't parsed correctly. I fixed it and added to test.
 
> For the second case, do you mean we should add a line if (open_bracket.ignore(pos, expected)) before L114.

You are right, see my commit.

> For the third case, it seems that we can allow both DISTINCT ON and DISTINCT?

Finally, decided to make this mutually exclusive.

**Review by vdimir — APPROVED** [review]
I made some changes for things that we discussed earlier, see my commits and some comments in resolved conversations.

**Comment by TszKitLo40** [comment]
> I made some changes for things that we discussed earlier, see my commits and some comments in resolved conversations.

Thanks for your commit. But it seems that some tests failed.

**Comment by vdimir** [comment]
@mergifyio update

**Comment by TszKitLo40** [comment]
@vdimir The distinct on test failed.

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+SELECT DISTINCT ON (a, b) * FROM t1;
+SELECT DISTINCT ON (a) * FROM t1;
+
+SELECT DISTINCT ON (a, b) a, b, c FROM t1 LIMIT 1 BY a, b; -- { clientError 590 }
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:10]:
```suggestion
SELECT DISTINCT ON (a, b) a, b, c FROM t1 LIMIT 1 BY a, b; -- { clientError 1 }
```

**Comment by vdimir** [comment]
`01268_procfs_metrics` failed, shouldn't be related, I'll try to investigate, but now we can update to master and run checks again.

@mergifyio update

**Comment by TszKitLo40** [comment]
Why the AST fuzzer test failed? @vdimir

File: src/Parsers/ParserSelectQuery.cpp [hunk 7]
```diff
@@ -335,6 +359,16 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
...
     }
 
+    if (distinct_on_expression_list)
+    {
+        /// DISTINCT ON and LIMIT BY are mutually exclusive, checked before
+        assert (limit_by_expression_list == nullptr);
+
+        /// Transform `DISTINCT ON expr` to `LIMIT 1 BY expr`
+        limit_by_expression_list = distinct_on_expression_list;
+        limit_by_length = std::make_shared<ASTLiteral>(Field{UInt8(1)});
```
Comment (vdimir) [inline src/Parsers/ParserSelectQuery.cpp:369]:
```suggestion
        limit_by_length = std::make_shared<ASTLiteral>(Field{UInt8(1)});
        distinct_on_expression_list = nullptr;
```

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 8]
```diff
@@ -0,0 +1,22 @@
+DROP TABLE IF EXISTS t1;
+
+CREATE TABLE t1 (`a` UInt32, `b` UInt32, `c` UInt32 ) ENGINE = Memory;
+INSERT INTO t1 VALUES (1, 1, 1), (1, 1, 2), (2, 2, 2), (1, 2, 2);
+
+SELECT DISTINCT ON (a, b) a, b, c FROM t1;
+SELECT DISTINCT ON (a, b) * FROM t1;
+SELECT DISTINCT ON (a) * FROM t1;
+
+SELECT DISTINCT ON (a, b) a, b, c FROM t1 LIMIT 1 BY a, b; -- { clientError 62 }
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:10]:
```suggestion
-- fuzzer will fail 
-- SELECT DISTINCT ON (a, b) a, b, c FROM t1 LIMIT 1 BY a, b; -- { clientError 62 }
```

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+SELECT DISTINCT ON a, b a, b FROM t1; -- { clientError 62 }
+SELECT DISTINCT ON a a, b FROM t1; -- { clientError 62 }
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:13]:
```suggestion
-- SELECT DISTINCT ON a, b a, b FROM t1; -- { clientError 62 }
-- SELECT DISTINCT ON a a, b FROM t1; -- { clientError 62 }
```

File: tests/queries/0_stateless/01917_distinct_on.sql [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+-- "Code: 47. DB::Exception: Missing columns: 'DISTINCT'" - error can be better
+SELECT DISTINCT ON (a, b) DISTINCT a, b FROM t1; -- { serverError 47 }
+SELECT DISTINCT DISTINCT ON (a, b) a, b FROM t1; -- { clientError 62 }
+
+SELECT ALL DISTINCT ON (a, b) a, b FROM t1; -- { clientError 62 }
+SELECT DISTINCT ON (a, b) ALL a, b FROM t1; -- { clientError 62 }
```
Comment (vdimir) [inline tests/queries/0_stateless/01917_distinct_on.sql:20]:
```suggestion
-- SELECT DISTINCT ON (a, b) DISTINCT a, b FROM t1; -- { serverError 47 }
-- SELECT DISTINCT DISTINCT ON (a, b) a, b FROM t1; -- { clientError 62 }

-- SELECT ALL DISTINCT ON (a, b) a, b FROM t1; -- { clientError 62 }
-- SELECT DISTINCT ON (a, b) ALL a, b FROM t1; -- { clientError 62 }
```

**Comment by vdimir** [comment]
I temporary removed queries with syntax error from test to merge this PR. When I figure out how not to cause fuzzer error with such queries, I'll enable it.

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11581.

**Pull request merged**